src/Terrain/MapInfo.java

```diff
@@ -40,4 +46,40 @@ public int getNumCos()
   {
     return COProperties.length;
   }
+
+  /**
+   * Determines which unit sets are valid for this map.
+   */
+  public UnitModelScheme[] getValidUnitModelSchemes()
+  {
+    ArrayList<UnitModelScheme> umsList = new ArrayList<UnitModelScheme>();
+    umsList.add(new AWBWUnits());
+    umsList.add(new DoRUnits());
+
+    // Filter based on the existence of non-core units in the map
+    for( UnitModelScheme scheme : umsList )
+    {
+      GameReadyModels models = scheme.getGameReadyModels();
+      boolean schemeInvalid = false;
+      
+      for( Map<XYCoord, String> unitSet : mapUnits )
+      {
+        for( Entry<XYCoord, String> unitEntry : unitSet.entrySet() )
+        {
+          UnitModel model = UnitModelScheme.getModelFromString(unitEntry.getValue(), models.unitModels);
+          if( model == null )
+          {
+            schemeInvalid = true; // This unit isn't supported in the scheme, so drop the scheme
+            break;
+          }
+        }
+        if( schemeInvalid )
+          break;
+      }
+
+      scheme.schemeValid = !schemeInvalid;
```

Comment:
Why not `scheme.schemeValid = false`? In theory, it will always be true when passed in, but still.

Reply:
They're different flags. One's local, and can be true or false depending on whether the map's units fit within the unit set.

Reply:
Right. Silly me.